File: android/app/src/main/java/com/currencyconverter/app/engine/ConverterEngine.kt
package com.currencyconverter.app.engine

import java.math.RoundingMode
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Locale
import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

/**
 * Pure currency-conversion math and text formatting.
 * No Android framework dependency -> fully unit testable on the JVM.
 *
 * Mirrors the reference JS implementation in
 * design/currency-converter-app-design/project/Currency Converter v2.dc.html
 */
object ConverterEngine {

    private val usSymbols = DecimalFormatSymbols(Locale.US)

    private fun groupedFormat(minFractionDigits: Int, maxFractionDigits: Int): DecimalFormat {
        val df = DecimalFormat()
        df.decimalFormatSymbols = usSymbols
        df.isGroupingUsed = true
        df.minimumFractionDigits = minFractionDigits
        df.maximumFractionDigits = maxFractionDigits
        df.roundingMode = RoundingMode.HALF_UP
        return df
    }

    /** result = amount / rate[from] * rate[to] */
    fun convert(amount: Double, fromRate: Double, toRate: Double): Double {
        if (fromRate == 0.0) return 0.0
        return amount / fromRate * toRate
    }

    /** "1 <FROM> = <fmt(rate[to]/rate[from])> <TO>" */
    fun rateLine(fromCode: String, toCode: String, fromRate: Double, toRate: Double): String {
        val perOne = if (fromRate == 0.0) 0.0 else toRate / fromRate
        return "1 $fromCode = ${fmt(perOne)} $toCode"
    }

    /** Display formatting for a computed/result value (not the raw entry buffer). */
    fun fmt(n: Double): String {
        val a = abs(n)
        return when {
            a >= 1e12 -> groupedFormat(0, 2).format(n / 1e12) + "T"
            a >= 1e9 -> groupedFormat(0, 2).format(n / 1e9) + "B"
            a >= 1e6 -> groupedFormat(0, 0).format(n)
            else -> {
                val maxDp = if (a >= 1) 2 else 4
                groupedFormat(2, maxDp).format(n)
            }
        }
    }

    /** Formats the live entry buffer (grouped integer part, raw decimal part as typed). */
    fun fmtEntry(s: String): String {
        val parts = s.split(".", limit = 2)
        val intPart = parts[0]
        val decPart = if (parts.size > 1) parts[1] else null
        val intVal = intPart.ifEmpty { "0" }.toDoubleOrNull() ?: 0.0
        val gi = groupedFormat(0, 0).format(intVal)
        return if (decPart == null) gi else "$gi.$decPart"
    }

    /** Dynamic font size (sp) for the big value text, per spec's `fit()`. */
    fun fit(s: String): Int {
        val len = max(s.length, 1)
        val computed = floor(300.0 / (0.62 * len)).toInt()
        return max(20, min(46, computed))
    }

    /** Strips grouping/formatting from a displayed string back into a raw numeric entry string. */
    fun stripFmt(s: String): String {
        val cleaned = s.filter { it.isDigit() || it == '.' }
        val n = cleaned.toDoubleOrNull() ?: 0.0
        val rounded = Math.round(n * 100.0) / 100.0
        return if (rounded == rounded.toLong().toDouble()) {
            rounded.toLong().toString()
        } else {
            var str = String.format(Locale.US, "%.2f", rounded)
            str = str.trimEnd('0').trimEnd('.')
            str
        }
    }

    /** Applies one keypad press to the raw entry buffer. Max 9 significant digits, single '.'. */
    fun press(entry: String, key: String): String {
        return when {
            key == "⌫" -> if (entry.length > 1) entry.dropLast(1) else "0"
            key == "." -> if (entry.contains(".")) entry else "$entry."
            entry.replace(".", "").length < 9 -> if (entry == "0") key else entry + key
            else -> entry
        }
    }
}

File: ios/CurrencyConverter/Engine/ConverterEngine.swift
import Foundation
import CoreGraphics

/// Pure, side-effect-free currency math + text formatting. Mirrors the reference
/// mock's `fmt` / `fmtEntry` / `fit` / `stripFmt` / `press` behaviour exactly
/// (see docs/SPEC.md section 2) so it can be unit tested independently of any UI.
enum ConverterEngine {

    // MARK: - Conversion

    /// `result = amount / rate[from] * rate[to]`
    static func convert(amount: Double, fromRate: Double, toRate: Double) -> Double {
        guard fromRate != 0 else { return 0 }
        return amount / fromRate * toRate
    }

    /// Parses the raw entry buffer (e.g. "1234.5") into a Double, matching
    /// `parseFloat(entry.replace(/,/g, '')) || 0`.
    static func parseEntry(_ entry: String) -> Double {
        let stripped = entry.replacingOccurrences(of: ",", with: "")
        return Double(stripped) ?? 0
    }

    /// `1 <fromCode> = <fmt(rate[to]/rate[from])> <toCode>`
    static func rateLineText(fromCode: String, toCode: String, fromRate: Double, toRate: Double) -> String {
        guard fromRate != 0 else { return "1 \(fromCode) = 0 \(toCode)" }
        return "1 \(fromCode) = \(fmt(toRate / fromRate)) \(toCode)"
    }

    // MARK: - Number formatting (`fmt`)

    /// Formats a numeric result for display, matching the mock's `fmt(n)`:
    /// - abs >= 1e12 -> value/1e12, up to 2dp, suffix "T"
    /// - abs >= 1e9  -> value/1e9,  up to 2dp, suffix "B"
    /// - abs >= 1e6  -> grouped integer, 0dp
    /// - else        -> grouped, min 2dp, max (abs>=1 ? 2 : 4)dp
    static func fmt(_ n: Double) -> String {
        let a = abs(n)
        if a >= 1e12 {
            return grouped(n / 1e12, minFrac: 0, maxFrac: 2) + "T"
        }
        if a >= 1e9 {
            return grouped(n / 1e9, minFrac: 0, maxFrac: 2) + "B"
        }
        if a >= 1e6 {
            return grouped(n, minFrac: 0, maxFrac: 0)
        }
        return grouped(n, minFrac: 2, maxFrac: a >= 1 ? 2 : 4)
    }

    /// Formats the raw keypad entry buffer for display while it is being typed:
    /// group the integer part with thousands separators, keep the decimal part
    /// (if any) exactly as typed. Matches the mock's `fmtEntry(s)`.
    static func fmtEntry(_ s: String) -> String {
        let parts = s.components(separatedBy: ".")
        let integerPart = parts[0]
        let integerValue = Double(integerPart.isEmpty ? "0" : integerPart) ?? 0
        let groupedInteger = grouped(integerValue, minFrac: 0, maxFrac: 0)
        guard parts.count > 1 else { return groupedInteger }
        return groupedInteger + "." + parts[1]
    }

    /// Strips a formatted display string back down to a plain numeric entry
    /// string rounded to 2dp, matching the mock's `stripFmt(s)`.
    static func stripFmt(_ s: String) -> String {
        let allowed = Set("0123456789.")
        let filtered = String(s.filter { allowed.contains($0) })
        let n = Double(filtered) ?? 0
        let rounded = (n * 100).rounded() / 100
        return plainNumberString(rounded)
    }

    // MARK: - Dynamic font size (`fit`)

    /// `max(20, min(46, floor(300 / (0.62 * max(len,1)))))`
    static func fit(_ s: String) -> CGFloat {
        let len = max(s.count, 1)
        let raw = floor(300.0 / (0.62 * Double(len)))
        let clamped = max(20.0, min(46.0, raw))
        return CGFloat(clamped)
    }

    // MARK: - Keypad entry buffer (`press`)

    /// Applies one keypad press to the entry buffer. Keys: "0"-"9", ".", "⌫".
    /// Max 9 significant digits (decimal point doesn't count). Leading "0" is
    /// replaced by the next digit. Single "." allowed. "⌫" deletes, floor is "0".
    static func press(_ key: String, entry: String) -> String {
        var e = entry
        if key == "⌫" {
            e = e.count > 1 ? String(e.dropLast()) : "0"
        } else if key == "." {
            if !e.contains(".") { e += "." }
        } else {
            let digitCount = e.filter { $0 != "." }.count
            if digitCount < 9 {
                e = (e == "0") ? key : e + key
            }
        }
        return e
    }

    // MARK: - Private helpers

    private static func grouped(_ n: Double, minFrac: Int, maxFrac: Int) -> String {
        let formatter = formatterCache.formatter(minFrac: minFrac, maxFrac: maxFrac)
        return formatter.string(from: NSNumber(value: n)) ?? String(n)
    }

    /// JS `String(n)` for a number already rounded to <=2dp: integers print
    /// with no decimal point, non-integers print trimmed (no trailing zeros).
    private static func plainNumberString(_ n: Double) -> String {
        if n == n.rounded(), abs(n) < 1e15 {
            return String(Int64(n))
        }
        var s = String(format: "%.2f", n)
        while s.hasSuffix("0") { s.removeLast() }
        if s.hasSuffix(".") { s.removeLast() }
        return s
    }

    /// Small cache of `NumberFormatter`s keyed by fraction-digit config, since
    /// constructing one is comparatively expensive and `fmt`/`fmtEntry` can be
    /// called many times per re-render while the user is typing.
    private final class FormatterCache: @unchecked Sendable {
        private var cache: [String: NumberFormatter] = [:]
        private let lock = NSLock()

        func formatter(minFrac: Int, maxFrac: Int) -> NumberFormatter {
            let key = "\(minFrac)-\(maxFrac)"
            lock.lock()
            defer { lock.unlock() }
            if let existing = cache[key] { return existing }
            let f = NumberFormatter()
            f.locale = Locale(identifier: "en_US")
            f.numberStyle = .decimal
            f.usesGroupingSeparator = true
            f.groupingSeparator = ","
            f.decimalSeparator = "."
            f.minimumFractionDigits = minFrac
            f.maximumFractionDigits = maxFrac
            cache[key] = f
            return f
        }
    }

    private static let formatterCache = FormatterCache()
}
